dvc/rwlock.py

```diff
@@ -0,0 +1,182 @@
+    ret = set(filter(path_info.overlaps, lock.keys()))
+    ret.add(path_info.url)
+    return ret
+
+
+def _check_no_writers(lock, info, path_infos):
+    for path_info in path_infos:
+        urls = _get_urls(lock["write"], path_info)
+        writers = [lock["write"].get(url) for url in urls]
+        writers = list(filter(bool, writers))
+        if not writers:
+            continue
+
+        assert len(writers) == 1
+        writer = writers[0]
+        if writer == info:
+            continue
+
+        raise LockError(
+            "'{}' is busy, it is being written to by:\n{}".format(
+                str(path_info), _info_to_str(writer)
+            )
+        )
+
+
+def _check_no_readers(lock, path_infos):
+    from itertools import chain
+
+    for path_info in path_infos:
+        urls = _get_urls(lock["read"], path_info)
+
+        readers = list(
+            chain.from_iterable([lock["read"].get(url, []) for url in urls])
+        )
+        if not readers:
+            continue
+
+        str_readers = "\n".join(_info_to_str(r) for r in readers)
+        raise LockError(
+            "'{}' is busy, it is being read by:\n{}".format(
+                str(path_info), str_readers
+            )
+        )
+
+
+def _acquire_read(lock, info, path_infos):
+    changes = []
+
+    for path_info in path_infos:
+        url = path_info.url
+
+        readers = lock["read"][url]
+        if info in readers:
+            continue
+
+        changes.append(url)
+        readers.append(info)
+
+    return changes
+
+
+def _acquire_write(lock, info, path_infos):
+    changes = []
+
+    for path_info in path_infos:
+        url = path_info.url
+
+        if lock["write"][url] == info:
+            continue
+
+        changes.append(url)
+        lock["write"][url] = info
+
+    return changes
+
+
+def _release_write(lock, info, changes):
+    for url in changes:
+        assert "write" in lock
+        assert url in lock["write"]
+        assert lock["write"][url] == info
+        del lock["write"][url]
+        if not lock["write"]:
+            del lock["write"]
+
+
+def _release_read(lock, info, changes):
+    for url in changes:
+        assert "read" in lock
+        assert url in lock["read"]
+        assert info in lock["read"][url]
+        lock["read"][url].remove(info)
+        if not lock["read"][url]:
+            del lock["read"][url]
+        if not lock["read"]:
+            del lock["read"]
+
+
+@contextmanager
+def rwlock(tmp_dir, cmd, read, write):
+    info = {"pid": os.getpid(), "cmd": cmd}
+
+    with _edit_rwlock(tmp_dir) as lock:
+        _check_no_writers(lock, info, read + write)
+        _check_no_readers(lock, write)
+
+        rchanges = _acquire_read(lock, info, read)
+        wchanges = _acquire_write(lock, info, write)
+
+    try:
+        yield
+    finally:
+        with _edit_rwlock(tmp_dir) as lock:
+            _release_write(lock, info, wchanges)
+            _release_read(lock, info, rchanges)
```

Comment:
So you made them reentrable, how will it work with threads? e.g. should we allow?

Another thing is that this r/wchanges logic only work until locks are either nested or non-overlapping, which kind of provided by context managers unless you hack them. If threads are involved locks might be overlapping non-nested though:

- func A acquires lock (changes)
- func B acquires lock (no changes)
- func A releases and removes the lock 
    - func B still relies on it being held
...
- func C comes in and does a mess

Reply:
Currently not needed, so these don't bother about threads at all. Don't think we need to add protection against threads here right now.

Reply:
I would comment that this is non-threadsafe and non-safe against overlapping but non-nested locks. This might appear at some point and we will forget about this limitation by that time.